fog/ledger/server/src/router_handlers.rs

```diff
@@ -0,0 +1,345 @@
+            Err(rpc_status) => return responses.fail(rpc_status).await,
+        }
+    }
+    responses.close().await?;
+    Ok(())
+}
+
+/// Handles a client's request by performing either an authentication or a
+/// query.
+pub async fn handle_request<E>(
+    mut request: LedgerRequest,
+    shard_clients: Vec<Arc<KeyImageStoreApiClient>>,
+    enclave: E,
+    logger: Logger,
+) -> Result<LedgerResponse, RpcStatus>
+where
+    E: LedgerEnclaveProxy,
+{
+    if request.has_auth() {
+        handle_auth_request(enclave, request.take_auth(), logger)
+    } else if request.has_check_key_images() {
+        handle_query_request(
+            request.take_check_key_images(),
+            enclave,
+            shard_clients,
+            logger,
+        )
+        .await
+        // TODO: Handle other cases here as they are added, such as the merkele
+        // proof service.
+    } else {
+        let rpc_status = rpc_invalid_arg_error(
+            "Inavlid LedgerRequest request",
+            "Neither the query nor auth fields were set".to_string(),
+            &logger,
+        );
+        Err(rpc_status)
+    }
+}
+
+/// The result of processing the MultiLedgerStoreQueryResponse from each Fog
+/// Ledger Shard.
+pub struct ProcessedShardResponseData {
+    /// gRPC clients for Shards that need to be retried for a successful
+    /// response.
+    pub shard_clients_for_retry: Vec<Arc<KeyImageStoreApiClient>>,
+
+    /// Uris for individual Fog Ledger Stores that need to be authenticated with
+    /// by the Fog Router. It should only have entries if
+    /// `shard_clients_for_retry` has entries.
+    pub store_uris_for_authentication: Vec<KeyImageStoreUri>,
+
+    /// New, successfully processed query responses.
+    pub new_query_responses: Vec<(ResponderId, attest::NonceMessage)>,
+}
+
+impl ProcessedShardResponseData {
+    pub fn new(
+        shard_clients_for_retry: Vec<Arc<KeyImageStoreApiClient>>,
+        store_uris_for_authentication: Vec<KeyImageStoreUri>,
+        new_query_responses: Vec<(ResponderId, attest::NonceMessage)>,
+    ) -> Self {
+        ProcessedShardResponseData {
+            shard_clients_for_retry,
+            store_uris_for_authentication,
+            new_query_responses,
+        }
+    }
+}
+
+/// Processes the MultiLedgerStoreQueryResponses returned by each Ledger Shard.
+pub fn process_shard_responses(
+    clients_and_responses: Vec<(Arc<KeyImageStoreApiClient>, MultiKeyImageStoreResponse)>,
+) -> Result<ProcessedShardResponseData, RouterServerError> {
+    let mut shard_clients_for_retry = Vec::new();
+    let mut store_uris_for_authentication = Vec::new();
+    let mut new_query_responses = Vec::new();
+
+    for (shard_client, mut response) in clients_and_responses {
+        // We did not receive a query_response for this shard.Therefore, we need to:
+        //  (a) retry the query
+        //  (b) authenticate with the Ledger Store that returned the decryption_error
```

Comment:
This comment might be better off shifted down so it's right above the match line for  `MultiKeyImageStoreResponseStatus::AUTHENTICATION_ERROR`, rather than deleted completely? It looks like this comment applies to that code-path.

Reply:
You're right, not sure how this ended up here. Will fix.